Remove Slicer-application specific stuff from Libs, Base and Loadable modules

AlexyPellegrini opened the pull request:
This is the first PR for the future SlicerLibrary support. This PR contains small changes that doesn't break anything in slicer and should be mergeable at any time.

This basically remove all hardcoded paths that only make sense for the Slicer application from the Base/, Libs /and Modules/Loadable/[Logic|MRML|MRMLDM|vtkWidgets] modules so they can used as libraries and still have the right behavior as long as the caller gives them the right informations. To do this, I removed all inclusions of `vtkSlicerConfiguration.h` from these modules!

This enables all these modules to be built the same way outside and inside the Slicer application, which will be important for SlicerLibrary! 

I want to integrate the changes one at a time to limit the size and complexity of the future SlicerLibrary Pull Request :) Next on the list is VTK 9.4 support.

Please review @jcfr @lassoan 
FYI: @Thibault-Pelletier @finetjul

lassoan commented:
Thank you, I'm very happy that we are beginning this work here! Overall the changes look good, there is only one thing that bothers me: using system environment variables for configuring the application. This could be quite fragile, as environment variables are global and very visible and can be changed by anyone at anytime in an uncontrollable manner. Could you think about a better, more controllable alternative for storing application configuration?

For example, we could have a json configuration file for simple key:value storage, implemented in a Slicer or MRML logic class (MRML is probably better, as there are some configuration options at MRML level and then we could use the same at both MRML and Slicer level). When the application is started, it would initialize some ApplicationConfiguration object from the config file at application startup.

AlexyPellegrini commented:
> Overall the changes look good, there is only one thing that bothers me: using system environment variables for configuring the application. This could be quite fragile, as environment variables are global and very visible and can be changed by anyone at anytime in an uncontrollable manner. Could you think about a better, more controllable alternative for storing application configuration?

I agree that environment variables are fragile, the first approach I tried was to add a member to vtkColorModuleLogic, and initialize it in qSlicerColorsModule::createLogic:
```cpp
  auto* logic = vtkSlicerColorLogic::New();
  qSlicerApplication * app = qSlicerApplication::application();
  logic->SetSlicerShareDirectory(app->slicerSharePath().toUtf8().toStdString());
```
Does this seem go to you?

lassoan commented:
Various module logic classes may need various application configuration information. Therefore, it would be simpler to just set the application logic object in the module logic object. Then the module logic object could obtain any information it needs from the application logic. I think we always set the application logic in module logic in the C++ Slicer application.

AlexyPellegrini commented on Base/Logic/vtkSlicerApplicationLogic.cxx:
```diff
@@ -711,26 +702,28 @@ bool vtkSlicerApplicationLogic::IsEmbeddedModule(const std::string& filePath,
 //----------------------------------------------------------------------------
 bool vtkSlicerApplicationLogic::IsPluginInstalled(const std::string& filePath,
-                                                  const std::string& applicationHomeDir)
+                                                  const std::string& applicationHomeDir,
```
these functions are static

lassoan replied:
I see. Then we have two options: either pass more variables (as implemented in this PR already) or make it a class method. However, considering that this method would now take 4 arguments that all come from the application logic object, making it a method of the application logic class may be more appropriate.

AlexyPellegrini replied:
I would require too much time, I will keep it this way.

AlexyPellegrini commented on Modules/Loadable/Colors/Logic/vtkSlicerColorLogic.cxx:
```diff
@@ -104,11 +101,20 @@ std::vector<std::string> vtkSlicerColorLogic::FindDefaultColorFiles()
     slicerHome = std::string(vtksys::SystemTools::GetEnv("SLICER_HOME"));
   }
+
+  const char *slicerShareDir = vtksys::SystemTools::GetEnv("SLICER_SHARE_DIR");
```
Done. I introduces more changes to properly integrate this change. I will let you review again.

AlexyPellegrini commented:
I removed all remaining references to the SLICER_HOME and SLICER_SHARE_DIR variables and put these information in vtkMRMLApplicationLogic instead. I also found out that vtkMRMLLogic class can now be removed, since it was only used for this.
I updated Colors modules initialization and tests to fix them.

AlexyPellegrini commented:
@lassoan @jcfr

Hi everyone,
Congratulations on the Slicer 5.8 release!

 I rebased this PR which is the first step for slicerlib support!

FYI @finetjul @Thibault-Pelletier

AlexyPellegrini commented:
@lassoan This is the first PR to look at if you got time!

jcfr commented:
Thanks for taking the time to put this pull request 🙏✨️

We will test locally and report back. Note that we may shuffle things around in term  of commit history.

finetjul reviewed (approved):
Minor comments, otherwise LGTM

AlexyPellegrini commented on Modules/Loadable/Colors/qSlicerColorsModule.cxx:
```diff
@@ -107,9 +107,11 @@ void qSlicerColorsModule::setup()
+  if (appLogic != nullptr)
   {
-    this->appLogic()->SetColorLogic(colorLogic);
+    colorLogic->SetMRMLApplicationLogic(appLogic);
```
it will never be, and would break existing code below if it could so I don't think that will be useful

finetjul reviewed (dismissed):
LGTM

lassoan commented:
We need the vtkMRMLLogic class. It contains useful utility functions that may be used by extensions and will be probably used in Slicer core. Some MRML library classes need application home and share folders. I agree that getting the value from environment variable was hacky, but it could get it instead from some singleton class that stores MRML-library-level settings (kind of like QSettings, but for MRML). Those settings could be set by the application at startup.

lassoan commented:
`vtkMRMLApplicationLogic` is not accessible from `MRML/Core`, so it is not a trivial change, but if we can avoid the need to access the home and share folders from `MRML/Core` then the moving should OK. Actually, it would be cleaner if MRML core does not need to know where these application folders are, so let's see if this is feasible without breaking things.

History: CLI modules and classes in `MRML/Core` used to rely on the `SLICER_HOME` environment variable to get the application home and share folders. We then introduced `vtkMRMLLogic` to centralize the mechanism of getting these folders (so that we can switch from using environment variables to something cleaner and more controllable). It is in `MRML/Core` to make sure it is accessible everywhere where MRML is used.

Current state: I've checked uses of the application home and share folders (via `vtkMRMLLogic` class and `SLICER_HOME`) and fortunately only a few uses left. Your PR changes Slicer core, so that's good. Use in FreeSurfer color node was changed some time ago to get the shared folder from the module logic. `CMFreg` extension still needs to access the home folder, but it uses the old `SLICER_HOME` mechanism, which will still work (and I think this extension is not used that much anyway), so it can be dealt with later. `SwissSkullStripper` uses `itksys::SystemTools::FindProgramPath` to find the CLI executable location and finds the module's shared data folder by relative path from that, which is not elegant, but workable (and usable for other CLI modules that need resource files).

Overall, the proposed change would not break anything now and it is cleaner design (less coupling) if CLI modules and MRML core classes do not access application settings or resource files, so let's proceed with moving of these methods to the MRML logic. If we later find the access is needed then we can implement a mechanism.

---

The scene cleanup helper functions are still useful, but of course we don't need to maintain a separate class for them. Probably the best new home would be the `vtkMRMLScene` class.

AlexyPellegrini commented:
@jcfr @lassoan Can we move this forward?

finetjul commented:
> @jcfr @lassoan Can we move this forward?

Isn't this PR blocked by https://github.com/Slicer/Slicer/issues/8477 ?

jcfr commented:
Thanks for the review, the pull request (https://github.com/Slicer/Slicer/pull/8644) extracting the relevant changes has been updated. I would appreciate if you could help get it finalized :pray:

jcfr commented on Base/QTCore/qSlicerUtils.cxx:
```diff
@@ -203,13 +204,27 @@ QString qSlicerUtils::extractModuleNameFromClassName(const QString& className)
+  std::string extensionsDirBase;
+#ifdef Slicer_BUILD_EXTENSIONMANAGER_SUPPORT
+  extensionsDirBase = Slicer_EXTENSIONS_DIRBASENAME;
+#endif
```
```suggestion
#ifdef Slicer_BUILD_EXTENSIONMANAGER_SUPPORT
  std::string extensionsDirBase = Slicer_EXTENSIONS_DIRBASENAME;
#else
  std::string extensionsDirBase;
#endif
```

jcfr commented on Modules/Loadable/Colors/Logic/Testing/Cxx/vtkSlicerColorLogicTest1.cxx:
```diff
@@ -26,6 +26,8 @@
 #include "vtkMRMLCoreTestingMacros.h"
 #include "vtkMRMLScene.h"
 
+#include "vtkSlicerConfigure.h"
```
```suggestion
#include "vtkSlicerConfigure.h" // For Slicer_SHARE_DIR
```

jcfr commented on Modules/Loadable/Colors/Logic/vtkSlicerColorLogic.cxx:
```diff
@@ -88,27 +88,28 @@ const char* vtkSlicerColorLogic::GetDefaultChartColorNodeID()
-    {
-      slicerHome = std::string("");
-    }
+    vtkErrorMacro("GetMRMLApplicationLogic() must not be null");
```
```suggestion
    vtkErrorMacro("FindDefaultColorFiles failed: Unable to get application logic");
```

jcfr commented on Modules/Loadable/Colors/Logic/vtkSlicerColorLogic.cxx:
```diff
@@ -88,27 +88,28 @@ const char* vtkSlicerColorLogic::GetDefaultChartColorNodeID()
+  if (shareDir.empty())
   {
-    slicerHome = std::string(vtksys::SystemTools::GetEnv("SLICER_HOME"));
+    vtkWarningMacro("ShareDirectory of vtkSlicerApplication is not defined.\n"
+                    "Default color files won't be loaded!");
```
```suggestion
    vtkWarningMacro("FindDefaultColorFiles failed: ShareDirectory is not set. "
                    "Default color files will not be loaded.");
```

jcfr commented:
As originally hinted by @lassoan in https://github.com/Slicer/Slicer/pull/8004#issuecomment-2432456593, storing settings in an object would be ideal. That said, this could definitely be done in a follow-up pull request.

After addressing those nitpicks, this will be good for integration :rocket: 

----

> [!NOTE]
> When configuring the corresponding slicer library to be distributed as Python wheels, we will have to further define the meaning of `IsPluginBuiltIn`.
